File: sandvich-android/app/src/main/java/io/github/mozilla/sandvich/rust/OAuthInfo.java
package io.github.mozilla.sandvich.rust;

import com.sun.jna.Pointer;
import com.sun.jna.Structure;

import java.util.Arrays;
import java.util.List;

public class OAuthInfo {
    /**
     * Represents a raw OAuthInfo pointer to a Rust struct.
     * Public for use with JNA; Raw should not be used in code beyond the FxA package.
     */
    /* package-local */
    public static class Raw extends Structure {
        public String access_token;
        public String keys;
        public String scope;

        public Raw(Pointer p) {
            super(p);
            read();
        }

        @Override
        protected List<String> getFieldOrder() {
            return Arrays.asList("access_token", "keys", "scope");
        }
    }

    public final String accessToken;
    public final String keys;
    public final String scope;

    OAuthInfo(Raw raw) {
        this.accessToken = raw.access_token;
        this.keys = raw.keys;
        this.scope = raw.scope;
        JNA.INSTANCE.fxa_oauth_info_free(raw.getPointer());
    }

}

File: fxa-client/sdks/android/library/src/main/java/org/mozilla/fxaclient/internal/OAuthInfo.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.fxaclient.internal

import com.sun.jna.Pointer
import com.sun.jna.Structure

import java.util.Arrays

class OAuthInfo internal constructor(raw: Raw) {

    val accessToken: String?
    val keys: String?
    val scope: String?

    class Raw(p: Pointer) : Structure(p) {
        @JvmField var accessToken: Pointer? = null
        @JvmField var keys: Pointer? = null
        @JvmField var scope: Pointer? = null

        init {
            read()
        }

        override fun getFieldOrder(): List<String> {
            return Arrays.asList("accessToken", "keys", "scope")
        }
    }

    init {
        try {
            this.accessToken = raw.accessToken?.getRustString()
            this.keys = raw.keys?.getRustString()
            this.scope = raw.scope?.getRustString()
        } finally {
            FxaClient.INSTANCE.fxa_oauth_info_free(raw.pointer)
        }
    }
}
